api/client/client_test.go

```diff
@@ -450,44 +462,278 @@ func TestRemoveProject(t *testing.T) {
+		}
+	})
+}
+
+func TestRegisterAppUser(t *testing.T) {
+	t.Parallel()
+	conf, client, done := setup(t)
+	defer done()
+
+	user := login(t, client, conf, "jon@doe.com")
+
+	t.Run("test register app user without token", func(t *testing.T) {
+		url := fmt.Sprintf("%s/register", conf.AddrGatewayUrl)
+		res, err := http.Post(url, "application/json", strings.NewReader("{}"))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer res.Body.Close()
+
+		var data map[string]string
+		if err := json.Unmarshal(body, &data); err != nil {
+			t.Fatal(err)
+		}
+		if _, ok := data["error"]; !ok {
+			t.Fatal("expected error in response body")
+		}
+		if res.StatusCode != http.StatusBadRequest {
+			t.Fatalf("expected status code 400, got %d", res.StatusCode)
+		}
+	})
+
+	project, err := client.AddProject(context.Background(), "foo", Auth{Token: user.SessionID})
+	if err != nil {
+		t.Fatal(err)
+	}
+	token, err := client.AddAppToken(context.Background(), project.ID, Auth{Token: user.SessionID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("test register app user", func(t *testing.T) {
+		req, err := json.Marshal(&map[string]string{
+			"token":     token.ID,
+			"device_id": uuid.New().String(),
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		url := fmt.Sprintf("%s/register", conf.AddrGatewayUrl)
+		res, err := http.Post(url, "application/json", bytes.NewReader(req))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer res.Body.Close()
+
+		var data map[string]string
+		if err := json.Unmarshal(body, &data); err != nil {
+			t.Fatal(err)
+		}
+		if e, ok := data["error"]; ok {
+			t.Fatalf("got error in response body: %s", e)
+		}
+		if id, ok := data["id"]; !ok {
+			t.Fatalf("response body missing id")
+		} else {
+			t.Logf("app user id: %s", id)
+		}
+		session, ok := data["session_id"]
+		if !ok {
+			t.Fatalf("response body missing session id")
+		} else {
+			t.Logf("app user session id: %s", session)
+		}
+		if res.StatusCode != http.StatusOK {
+			t.Fatalf("expected status code 200, got %d", res.StatusCode)
+		}
+
+		// Setup a threads client for app user tests.
+		ttarget, err := tutil.TCPAddrFromMultiAddr(conf.AddrThreadsApi)
+		if err != nil {
+			t.Fatal(err)
+		}
+		tclient, err := threadsclient.NewClient(ttarget, grpc.WithInsecure(),
+			grpc.WithPerRPCCredentials(collections.TokenAuth{}))
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer tclient.Close()
+
+		t.Run("test create new store without client token", func(t *testing.T) {
+			if _, err := tclient.NewStore(context.Background()); err == nil {
+				t.Fatal("test create new store without client token should fail")
+			}
+		})
+
+		t.Run("test create new store with client token", func(t *testing.T) {
+			storeID, err := tclient.NewStore(collections.AuthCtx(context.Background(), session))
```

Comment:
Here, an app user is using a session from registration to create their own store.

Reply:
Cool. What does it mean to have a user store? Is it somehow restricted access to just that user?